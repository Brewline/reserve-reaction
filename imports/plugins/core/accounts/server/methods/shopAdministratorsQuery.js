import { Meteor } from "meteor/meteor";

/**
 * @name shopAdministratorsQuery
 * @method
 * @memberof Accounts/NoMeteorQueries
 * @summary return Account object for all users who are "owner" or "admin" role for the shop
 * @param {Object} context - an object containing the per-request state
 * @param {String} id - ID of shop
 * @return {Object[]} Array of user account objects
 */
export async function shopAdministratorsQuery(context, id) {
  const { collections, userHasPermission } = context;
  const { Accounts, users: Users } = collections;

  if (!userHasPermission(["owner", "admin"], id)) throw new Meteor.Error("access-denied", "User does not have permission");

  const users = await Users.find({
    [`roles.${id}`]: "admin"
  }, {
    fields: { _id: 1 }
  }).toArray();
  const userIds = users.map(({ _id }) => _id);

  return Accounts.find({ _id: { $in: userIds } });
}
