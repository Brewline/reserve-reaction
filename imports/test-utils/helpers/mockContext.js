const mockContext = {
  collections: {},
  shopId: "FAKE_SHOP_ID",
  userHasPermission: jest.fn().mockName("userHasPermission"),
  userId: "FAKE_USER_ID"
};

[
  "Accounts",
  "Assets",
  "Cart",
  "Catalog",
  "Emails",
  "Groups",
  "Inventory",
  "Logs",
  "MediaRecords",
  "Notifications",
  "Orders",
  "Packages",
  "Products",
  "Revisions",
  "roles",
  "SellerShops",
  "Shipping",
  "Shops",
  "Sms",
  "Tags",
  "Templates",
  "Themes",
  "Translations",
  "users"
].forEach((collectionName) => {
  mockContext.collections[collectionName] = {
    insert() {
      throw new Error("insert mongo method is deprecated, use insertOne or insertMany");
    },
    remove() {
      throw new Error("remove mongo method is deprecated, use deleteOne or deleteMany");
    },
    update() {
      throw new Error("update mongo method is deprecated, use updateOne or updateMany");
    },
    deleteOne: jest.fn().mockName(`${collectionName}.deleteOne`),
    deleteMany: jest.fn().mockName(`${collectionName}.deleteMany`),
    find: jest
      .fn()
      .mockName(`${collectionName}.find`)
      .mockReturnThis(),
    findOne: jest.fn().mockName(`${collectionName}.findOne`),
    insertOne: jest.fn().mockName(`${collectionName}.insertOne`),
    insertMany: jest.fn().mockName(`${collectionName}.insertMany`),
    toArray: jest.fn().mockName(`${collectionName}.toArray`),
    updateOne: jest.fn().mockName(`${collectionName}.updateOne`),
    updateMany: jest.fn().mockName(`${collectionName}.updateMany`)
  };
});

mockContext.collections.Media = {
  find: jest.fn().mockName("Media.find"),
  findLocal: jest.fn().mockName("Media.findLocal"),
  findOne: jest.fn().mockName("Media.findOne"),
  findOneLocal: jest.fn().mockName("Media.findOneLocal"),
  insert: jest.fn().mockName("Media.insert"),
  update: jest.fn().mockName("Media.update"),
  remove: jest.fn().mockName("Media.remove")
};

export default mockContext;
