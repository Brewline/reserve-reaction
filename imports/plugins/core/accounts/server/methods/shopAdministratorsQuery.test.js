import mockContext from "/imports/test-utils/helpers/mockContext";
import { shopAdministratorsQuery } from "./shopAdministratorsQuery";

const fakeShopId = "FAKE_SHOP_ID_ADMIN_QUERY";

beforeEach(() => {
  jest.resetAllMocks();
});

test("throws if userHasPermission returns false", async () => {
  mockContext.userHasPermission.mockReturnValueOnce(false);
  await expect(shopAdministratorsQuery(mockContext, fakeShopId)).rejects.toThrowErrorMatchingSnapshot();
  expect(mockContext.userHasPermission).toHaveBeenCalledWith(["owner", "admin"], fakeShopId);
});

test("returns accounts cursor if user has permission", async () => {
  mockContext.userHasPermission.mockReturnValueOnce(true);
  mockContext.collections.users.find.mockReturnValueOnce({
    toArray: () => Promise.resolve([{ _id: "user1" }, { _id: "user2" }])
  });
  mockContext.collections.Accounts.find.mockReturnValueOnce("CURSOR");
  const result = await shopAdministratorsQuery(mockContext, fakeShopId);
  expect(mockContext.userHasPermission).toHaveBeenCalledWith(["owner", "admin"], fakeShopId);
  expect(mockContext.collections.Accounts.find).toHaveBeenCalledWith({ _id: { $in: ["user1", "user2"] } });
  expect(result).toBe("CURSOR");
});
