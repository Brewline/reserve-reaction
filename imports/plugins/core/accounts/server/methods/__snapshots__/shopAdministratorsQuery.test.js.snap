// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`throws if userHasPermission returns false 1`] = `"access-denied"`;
